src/System.Management.Automation/engine/hostifaces/LocalPipeline.cs

```diff
@@ -1297,6 +1297,10 @@ public void Dispose()
     /// </summary>
     internal class PipelineStopper
     {
+        private readonly SemaphoreSlim _criticalSectionSemaphore = new SemaphoreSlim(1, 1);
```

Comment:
Problem is that locks re-enter based on thread.  This implementation basically does re-entrance by task.  So using threads as a metaphor, the pipeline stopper is one thread and critical regions are another.

Reply:
So the implication here is that we have multiple threads passing around the same essentially sequential execution, right? We're not using any async APIs here, so I'm assuming this is due to PowerShell's own passing around of executions between threads. 

Reading through the lock-related code, we basically have a situation where we want to stop the command running, but need to ensure that we're not disposing it at the same time.

We could try a recursive `ReaderWriterLock` for that, where the disposer is the writer (it's writing state to the command) and the command processor is the reader.

Reply:
> So the implication here is that we have multiple threads passing around the same essentially sequential execution, right? We're not using any async APIs here, so I'm assuming this is due to PowerShell's own passing around of executions between threads.

Yeah.  iirc an example is in remoting scenarios sometimes `Dispose` is called from a different thread, but the pipeline is still synchronously waiting for it's completion.  That sound right @vexx32?

> Reading through the lock-related code, we basically have a situation where we want to stop the command running, but need to ensure that we're not disposing it at the same time.

This might be paraphrasing what you said, but the main thing is to stop dispose related code itself from being stopped. So basically hold the pipeline stopper while `cleanup` is running.

Reply:
Yeah that sounds about right. The other easy example is calling `Stop()` on a `PowerShell` instance from another thread while it's currently processing something, iirc.